### PR TITLE
[Ruby] Add support for Twirp framework

### PR DESCRIPTION
**Pull request by pwntester**
Support for [Twirp framework](https://twitchtv.github.io/twirp/):
- Sources of untrusted data
- SSRF sinks

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/frameworks/Twirp.qll [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * Gets a local source node for the Service instantiation argument (the service handler).
+     */
+    DataFlow::LocalSourceNode getHandlerSource() { result = this.getArgument(0).getALocalSource() }
+
+    /**
+     * Gets the API::Node for the service handler's class.
+     */
+    API::Node getHandlerClassApiNode() { result.getAnInstantiation() = this.getHandlerSource() }
```
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/frameworks/Twirp.qll:33]: Probably `getAHandlerClassApiNode()` as there may be multiple handler sources.

File: ruby/ql/lib/codeql/ruby/frameworks/Twirp.qll [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+     * Gets the local source node for the service handler's class.
+     */
+    DataFlow::LocalSourceNode getHandlerClassDataFlowNode() {
+      result = this.getHandlerClassApiNode().asSource()
+    }
+
+    /**
+     * Gets the AST module for the service handler's class.
+     */
+    Ast::Module getHandlerClassAstNode() {
```
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/frameworks/Twirp.qll:45]: Suggest `getAHandlerClassAstNode()`

File: ruby/ql/lib/codeql/ruby/frameworks/Twirp.qll [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+private import codeql.ruby.Concepts
+
+/**
+ * Provides classes for modeling the `Twirp` framework.
+ */
+module Twirp {
+  /**
+   * A Twirp service instantiation
+   */
+  class ServiceInstantiation extends DataFlow::CallNode {
```
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/frameworks/Twirp.qll:19]: A lot of the predicates in this class feel like implementation details - I'd suggest making all of them `private` except for `getHandlerMethod()` as we're committing to maintaining any public methods as part of the public API.

File: ruby/ql/lib/change-notes/2023-02-03-twirp.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+ category: minorAnalysis
+---
+* Support for [Twirp framework](https://twitchtv.github.io/twirp/docs/intro.html).
```
Comment (alexrford) [inline ruby/ql/lib/change-notes/2023-02-03-twirp.md:4]: We should probably add a few more words to describe which parts of Twirp are modelled here.

File: ruby/ql/test/library-tests/frameworks/Twirp/tests.expected [hunk 5]
```diff
@@ -0,0 +1 @@
+The query depends on an extensional predicate sinkModel which has not been defined.
```
Comment (alexrford) [inline ruby/ql/test/library-tests/frameworks/Twirp/tests.expected:1]: This file needs to be deleted.

**Review by alexrford — CHANGES_REQUESTED** [review]
Sorry for the delay in reviewing this. Mostly style related comments, but this seems to work as expected otherwise. The tests are currently failing due to a missing `Twirp.expected` output.

File: ruby/ql/lib/codeql/ruby/frameworks/Twirp.qll (hunk 3 above)
Comment (pwntester) [reply ruby/ql/lib/codeql/ruby/frameworks/Twirp.qll:19]: Will do, thanks!

File: ruby/ql/test/library-tests/frameworks/Twirp/Twirp.ql [hunk 6]
```diff
@@ -0,0 +1,10 @@
+private import codeql.ruby.frameworks.Twirp
```
[the remark this answers is not in the document]
Comment (pwntester) [reply ruby/ql/test/library-tests/frameworks/Twirp/Twirp.ql:1]: I tried that, but Im getting this error `The query depends on an extensional predicate sinkModel which has not been defined.` and an empty actual file:

```
┗╸❯❯❯ codeql test run ruby/ql/test/library-tests/frameworks/Twirp 
Executing 1 tests in 1 directories.
Extracting test database in /Users/pwntester/src/github.com/github/codeql/ruby/ql/test/library-tests/frameworks/Twirp.
Compiling queries in /Users/pwntester/src/github.com/github/codeql/ruby/ql/test/library-tests/frameworks/Twirp.
Executing tests in /Users/pwntester/src/github.com/github/codeql/ruby/ql/test/library-tests/frameworks/Twirp.
Cannot find Twirp.expected file.
--- expected
+++ actual
@@ -1,1 +1,1 @@
-
+The query depends on an extensional predicate sinkModel which has not been defined.
[1/1 comp 6.3s eval 2ms] FAILED(EXECUTION) /Users/pwntester/src/github.com/github/codeql/ruby/ql/test/library-tests/frameworks/Twirp/Twirp.ql
0 tests passed; 1 tests failed:
  FAILED: /Users/pwntester/src/github.com/github/codeql/ruby/ql/test/library-tests/frameworks/Twirp/Twirp.ql
```

[the remark this answers is not in the document]
Comment (pwntester) [reply ruby/ql/test/library-tests/frameworks/Twirp/Twirp.ql:1]: Does it work on your end?

[the remark this answers is not in the document]
Comment (alexrford) [reply ruby/ql/test/library-tests/frameworks/Twirp/Twirp.ql:1]: The tests run on my machine (CodeQL 2.12.2) - I can't say exactly why they're failing for you locally, but its likely got to do with your CodeQL version or config.

[the remark this answers is not in the document]
Comment (pwntester) [reply ruby/ql/test/library-tests/frameworks/Twirp/Twirp.ql:1]: Thanks!

**Review by alexrford — APPROVED** [review]
Ran a quick performance check and everything looks fine.

**Pull request merged**